File: 8-Taller2DeSQL-CreandoUnCRUD/contacts-backend/src/main/java/com/contacts/contactsbackend/service/ContactService.java
package com.contacts.contactsbackend.service;

import com.contacts.contactsbackend.dao.ContactDao ;
import com.contacts.contactsbackend.domain.Contact ;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.Optional;

/**
 * @author Pablo Hernán Gatti
 */

@Service
public class ContactService implements IContactService{
    /**
     * Inyecta ContactDao
     */
    @Autowired
    private ContactDao contactDao;

    /**
     * Lista los contactos mientras DeletedAt sea null
     */
    @Override
    @Transactional(readOnly = true)
    public List<Contact> findByDeletedAtIsNull() {
        return contactDao.findByDeletedAtIsNull();
    }
    /**
     * Crea un contacto
     */
    //@Override
    @Transactional
    public Contact save(Contact contact) {
        return contactDao.save(contact);
    }
    /**
     * Realiza el borrado lógico de un contacto
     */
    @Transactional
    public void deleteLogical(Long id, Contact contact){
        contactDao.deleteLogical(id, contact.getDeletedAt());
    }
    /**
     * Realiza el borrado físico de un contacto
     */
    @Override
    @Transactional
    public void delete(Long id) {
        contactDao.delete(id);
    }
    /**
     * Busca un contacto por id
     */
    @Override
    public Optional<Contact> findContact(Long id) {
        return contactDao.findById(id);
    }
    /**
     * Actualiza un contacto
     */
    //@Override
    @Transactional
    public Contact update(Long id, Contact contact) {
        contact.setId(id);
        return contactDao.save(contact);
    }
}

File: 8-Taller2DeSQL-CreandoUnCRUD/contacts-frontend/src/app/contact.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Contact } from './contact';

/**
 * @author Pablo Hernán Gatti
 */

@Injectable({
  providedIn: 'root',
})
export class ContactService {
  /**
   * Ruta donde se encuentra el backend
   */
  private url = 'http://localhost:8080/api/contacts';

  /**
   * Creación de fecha para realizar el borrado lógico
   */
  date = Date.now();

  constructor(private httpClient: HttpClient) {}

  /**
   * Lista los contactos mientras el atributo "deletedAt" sea nulo
   * @returns Devuelve la lista de contactos desde el backend
   */
  getContactsList(): Observable<Contact[]> {
    return this.httpClient.get<Contact[]>(`${this.url}`);
  }
  /**
   * Crea un contacto nuevo
   * @param contact recibe un objeto Contacto
   * @returns Devuelve un contacto nuevo desde el backend
   */
  addContact(contact: Contact): Observable<Object> {
    return this.httpClient.post(`${this.url}`, contact);
  }

  /**
   * Borrado físico
   * @param id Recibe el id de tipo number para identificar el contacto a eliminar
   * @returns Devuelve el borrado desde el backend
   */
  deleteContact(id: number): Observable<Object> {
    return this.httpClient.delete(`${this.url}/${id}`);
  }

  /**
   * Borrado lógico: Añade un timestamp al contacto en el atributo "deletedAt" para que no se muestre en la lista
   * @param id Recibe el id de tipo number para identificar el contacto a eliminar
   * @returns Devuelve el borrado desde el backend
   */
  safeDeleteContact(id: number): Observable<Object> {
    return this.httpClient.patch(`${this.url}/${id}`, {
      deletedAt: this.date,
    });
  }
  /**
   * Busca y devuelve un contacto según el id
   * @param id Recibe el id de tipo number para identificar el contacto a buscar
   * @returns Devuelve el contacto perteneciente al id ingresado desde el backend
   */
  findContact(id: number): Observable<Contact> {
    return this.httpClient.get<Contact>(`${this.url}/${id}`);
  }
  /**
   * Actualiza un contacto
   * @param id Recibe el id de tipo number para identificar el contacto a actualizar
   * @param contact Recibe el contacto a actualizar
   * @returns Devuelve el contacto perteneciente al id ingresado desde el backend
   */
  updateContact(id: number, contact: Contact): Observable<Object> {
    return this.httpClient.put(`${this.url}/${id}`, contact);
  }
}
